pkg/webhook/authorization/webhook.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authorization
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	authorizationv1 "k8s.io/api/authorization/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+var (
+	errUnableToEncodeResponse = errors.New("unable to encode response")
+)
+
+// Request defines the input for an authorization handler.
+// It contains information to identify the object in
+// question (group, version, kind, resource, subresource,
+// name, namespace), as well as the operation in question
+// (e.g. Get, Create, etc), and the object itself.
+type Request struct {
+	authorizationv1.SubjectAccessReview
+}
+
+// Response is the output of an authorization handler.
+// It contains a response indicating if a given
+// operation is allowed
+type Response struct {
+	authorizationv1.SubjectAccessReview
+}
+
+// Complete populates any fields that are yet to be set in
+// the underlying TokenResponse, It mutates the response.
+func (r *Response) Complete(req Request) error {
+	r.UID = req.UID
+
+	return nil
+}
+
+// Handler can handle an SubjectAccessReview.
+type Handler interface {
+	// Handle yields a response to an SubjectAccessReview.
+	//
+	// The supplied context is extracted from the received http.Request, allowing wrapping
+	// http.Handlers to inject values into and control cancelation of downstream request processing.
+	Handle(context.Context, Request) Response
+}
```

Comment:
In Gardener ([ref](https://github.com/gardener/gardener/blob/7c4388ee8bb0eaecfbae451772dc2577df45cc3f/pkg/admissioncontroller/webhooks/auth/seed/handler.go#L63)), the interface for the handler is [`k8s.io/apiserver/pkg/authorization/authorizer.Authorizer`](https://github.com/kubernetes/apiserver/blob/c13376c1a6cd7f67c1b72f1f96256164d6b592f2/pkg/authorization/authorizer/interfaces.go#L67-L72), i.e.

```golang
type Authorizer interface {
	Authorize(ctx context.Context, a Attributes) (authorized Decision, reason string, err error)
}
```

The reason was that this is also the interface for the in-tree authorizers in k/k.

Do you think it makes sense to also change the `Request` type in this PR to inline `auth.Attributes` instead of the plain `SubjectAccessReview` type?

Reply:
I personally think it's nice to align them with the rest of controller-runtime webhooks like you've done but @DirectXMan12 might have thoughts here.